custom/src/main/java/com/splunk/opentelemetry/EndpointProtocolValidator.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry;
+
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+import java.util.function.Consumer;
+
+class EndpointProtocolValidator {
+
+  private final Map<String, String> customized;
+  private final ConfigProperties config;
+  private final Consumer<String> warn;
+
+  private EndpointProtocolValidator(
+      Map<String, String> customized, ConfigProperties config, Consumer<String> warn) {
+    this.customized = customized;
+    this.config = config;
+    this.warn = warn;
+  }
+
+  /** Warn when the port number in OTLP endpoint doesn't agree with configured protocol. */
+  static void validate(
+      Map<String, String> customized, ConfigProperties config, Consumer<String> warn) {
+    EndpointProtocolValidator validator = new EndpointProtocolValidator(customized, config, warn);
+    validator.verifyEndpointProtocol();
+  }
+
+  private void verifyEndpointProtocol() {
+    String protocol = getString("otel.exporter.otlp.protocol", "http/protobuf");
+    String endpoint =
+        getString(
+            "otel.exporter.otlp.endpoint",
+            "http://localhost:" + ("http/protobuf".equals(protocol) ? "4318" : "4317"));
+    if (!verifyEndpointProtocol(
+        protocol, endpoint, "otel.exporter.otlp.protocol", "otel.exporter.otlp.endpoint")) {
+      // there already was a mismatch between endpoint port and protocol,
+      // skip looking at signal specific endpoints
+      return;
+    }
+
+    for (String signal : new String[] {"traces", "metrics", "logs"}) {
+      verifySignalEndpointProtocol(signal, protocol, endpoint);
+    }
+  }
+
+  private boolean verifyEndpointProtocol(
+      String protocol, String endpoint, String protocolKey, String endpointKey) {
+    int port = getEndpointPort(endpoint);
+    if ("http/protobuf".equals(protocol)) {
+      if (port == 4317) {
```

Comment:
Perhaps `String.valueOf(port).equals(getPort("grpc"))` and same for `"http/protobuf"` in the next block? Some extra work, but would only have one instance of 4318 and 4317 literals in this file then, plus it is clearer what this checks for when you don't remember which port number was for which protocol.